Hash sample optimization

glubsy opened the pull request:
For files bigger than the threshold set by the user, instead of computing the full md5 (which takes time), pick three chunks located at 25%, 60% of the file length, then the last chunk, and compute this "md5samples" hash instead from these three samples. 

Closes #734.

glubsy committed:
Add partial hashes optimization for big files

* Big files above the user selected threshold can be partially hashed in 3 places.
* If the user is willing to take the risk, we consider files with identical md5samples as being identical.

glubsy committed:
Add unit tests for hash sample optimization

* Instead of keeping md5 samples separate, merge them as one hash computed from the various selected chunks we picked.
* We don't need to keep a boolean to see whether or not the user chose to optimize; we can simply compare the value of the threshold, since 0 means no optimization currently active.

glubsy commented:
Here is what it looks like in the preference dialog window: 
![2021-06-21_22-42-27](https://user-images.githubusercontent.com/12895548/122826585-a1451380-d2e3-11eb-8c2a-a60f95abfb3f.png)

glubsy committed:
Remove unused import

arsenetar commented:
Preliminary review looks good, I'll need to pull this and give it a try.

arsenetar commented:
@glubsy, did some testing with this and I am wondering if we should force a limit on the minimum value here (currently allows 1MB).  The reason being the CHUNK_SIZE used for the three hashes is 1MB so if you set the limit to anything smaller than 3MB you are actually doing more work than just hashing the whole file once.  I don't know why you would actively set the limit that low, but one could, I tested this with the limit at 1MB and it does not cause issues but does do more work than just hashing the files I tested with (just over 1MB).

arsenetar commented:
Ran the tests locally while they pass, the negative seek is causing an exception which is caught on test cases which have files smaller than the chunk size.  I would prefer either the test be updated to test files that are 1MB or larger (as you cannot actually specify for it to run on smaller files), or write that seek as `fp.seek(-min(size, CHUNK_SIZE),2)`.

glubsy commented:
That's a very good point, thanks! I'll work on this very shortly.

ElDavoo commented:
Saved my life in analyzing my movie collection, thanks!

glubsy commented:
Had a quick look today, and it just occurred to me:

```python3
if size <= CHUNK_SIZE * 3:
    setattr(self, field, self.md5)
```
Wouldn't this be enough?

glubsy committed:
Avoid partially hashing small files

Computing 3 hash samples for files less than 3MiB (3 * CHUNK_SIZE) is not efficient since spans of later samples would overlap a previous one.
Therefore we can simply return the hash of the entire small file instead.

glubsy committed:
Fix for older python versions

The "walrus" operator is only available in python 3.8 and later. Fall back to more traditional notation.

glubsy committed:
Cache constant expression

Perhaps the python byte code is already optimized, but just in case it is not, keep pre-compute the constant expression.

glubsy committed:
Fix flake 8

arsenetar commented:
Yeah that looks good to me, tests seem to be good now as well. I am thinking of changing the edit box for this value and the ignore small files to spinboxes instead to prevent some other input validation issues unrelated to this change so I'll handle that separately from this PR.  Thanks for the work.
